### PR TITLE
Refactoring change: IsLoopFusion() etc.

### PR DESCRIPTION
**Pull request by sana-damani**
Create IsLoopFusion(), IsInputFusion(), IsOutputFusion() and IsCustomFusion() methods.

Tested using //tensorflow/compiler/xla/service/gpu/tests:gpu_fusion_test and //tensorflow/python/compiler

### CODE REVIEW

**Comment by rthadur** [comment]
@sana-damani thanks for your contribution , is it possible to combine this PR #26301 with this.

**Comment by sana-damani** [comment]
> @sana-damani thanks for your contribution , is it possible to combine this PR #26301 with this.

I'd like to keep independent changes separate unless you think there is some benefit in combining the two.

**Comment by rthadur** [comment]
> > @sana-damani thanks for your contribution , is it possible to combine this PR #26301 with this.
> 
> I'd like to keep independent changes separate unless you think there is some benefit in combining the two.

@sana-damani combining helps to fasten the process of review and merging internally, let me know

**Comment by thomasjoerg** [comment]
> > > @sana-damani thanks for your contribution , is it possible to combine this PR #26301 with this.
> > 
> > 
> > I'd like to keep independent changes separate unless you think there is some benefit in combining the two.
> 
> @sana-damani combining helps to fasten the process of review and merging internally, let me know

@rthadur  Hi, I've been working with Sana and recommended to keep PRs small and focused. I'm doing the internal code reviews and prefer this approach. Does it create overhead for you when triaging PRs, for instance? If so, is there anything we could do to avoid this? I'd like to make sure the process works smoothly for all of us.

File: tensorflow/compiler/xla/service/gpu/multi_output_fusion.cc [hunk 1]
```diff
@@ -155,9 +152,7 @@ bool GpuMultiOutputFusion::DoProducerConsumerMultiOutputFusion() {
         VLOG(3) << producer->name() << " is a constant.";
         continue;
       }
-      const bool is_loop_fusion =
-          producer->opcode() == HloOpcode::kFusion &&
-          producer->fusion_kind() == HloInstruction::FusionKind::kLoop;
+      const bool is_loop_fusion = producer->IsLoopFusion();
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/multi_output_fusion.cc]: Nit: Please inline this.

**Review by thomasjoerg — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/multi_output_fusion.cc (hunk 1 above)
Comment (sana-damani) [reply tensorflow/compiler/xla/service/gpu/multi_output_fusion.cc]: Would you like me to turn "IsLoopFusion()" into an inline function? Or replace is_loop_fusion with calls to IsLoopFusion()?

Comment (thomasjoerg) [reply tensorflow/compiler/xla/service/gpu/multi_output_fusion.cc]: My idea was to replace line 156 with
`if (!producer->IsElementwise() &&  !producer->IsLoopFusion())`

**Review by thomasjoerg — DISMISSED** [review]

**Comment by rthadur** [comment]
> > > > @sana-damani thanks for your contribution , is it possible to combine this PR #26301 with this.
> > > 
> > > 
> > > I'd like to keep independent changes separate unless you think there is some benefit in combining the two.
> > 
> > 
> > @sana-damani combining helps to fasten the process of review and merging internally, let me know
> 
> @rthadur Hi, I've been working with Sana and recommended to keep PRs small and focused. I'm doing the internal code reviews and prefer this approach. Does it create overhead for you when triaging PRs, for instance? If so, is there anything we could do to avoid this? I'd like to make sure the process works smoothly for all of us.

@thomasjoerg no issues , i was trying to combine related PR so that it would be easy for everyone. @sana-damani could you please rebase the branch

**Comment by sana-damani** [comment]
I had to resolve some conflicts and merge new changes.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer : You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/compiler/xla/service/gpu/multi_output_fusion.cc [hunk 2]
```diff
@@ -47,13 +47,8 @@ bool GpuMultiOutputFusion::ShapesCompatibleForFusion(HloInstruction* instr1,
...
   // We can fuse reduces and loop fusions. Elementwise instructions can be fused
   // with any other instruction.
-  // TODO(b/112957171): This should use the same isFusible logic as
-  // instruction_fusion.
   return instr->IsFusible() &&
-         (IsInputFusibleReduction(*instr) ||
-          (instr->opcode() == HloOpcode::kFusion &&
-           instr->fusion_kind() == HloInstruction::FusionKind::kLoop) ||
-          instr->IsElementwise());
+         (IsInputFusibleReduction(*instr) || instr->IsLoopFusion());
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/multi_output_fusion.cc]: This change makes the fusion criteria more restrictive: Unfused element-wise ops are no longer considered fusible.
I probably confused you with my earlier comment. My idea was to call IsLoopFusible instead, which does allow unfused, element-wise ops. However, thinking about this again, it also allows certain non-elmentwise ops. While multi-output fusion should handle them just fine, it's probably a good idea to keep this PR a pure refactoring without any functional changes.
In summary, please keep the 
`|| instr->IsElementwise()`

**Review by thomasjoerg — DISMISSED** [review]
Approving this for merging. Please address my comment so we can start the internal review.
